### PR TITLE
feat: provide redis shared memory plugin

### PR DESCRIPTION
**Pull request by n063h**
provide RedisShmWrapper class and an example.

### CODE REVIEW

File: docs/source/examples/ipc.md [hunk 1]
```diff
@@ -4,7 +4,7 @@ This is an example demonstrating how you can enable the plasma shared memory sto
 
 Mosec's multi-stage pipeline requires the output data from the previous stage to be transferred to the next stage across python processes. This is coordinated via Unix domain socket between every Python worker process from all stages and the Rust controller process.
 
-By default, we serialize the data and directly transfer the bytes over the socket. However, users may find wrapping this IPC useful or more efficient for specific use cases. Therefore, we provide the `mosec.plugins.IPCWrapper` interface and an example implementation `PlasmaShmWrapper` based on [`pyarrow.plasma`](https://arrow.apache.org/docs/python/plasma.html).
+By default, we serialize the data and directly transfer the bytes over the socket. However, users may find wrapping this IPC useful or more efficient for specific use cases. Therefore, we provide the `mosec.plugins.IPCWrapper` interface and example implementation `PlasmaShmWrapper` based on [`pyarrow.plasma`](https://arrow.apache.org/docs/python/plasma.html) and `RedisShmWrapper` based on [`redis`](https://pypi.org/project/redis). We recommand using `RedisShmWrapper` for better performance.
```
Comment (lkevinzc) [inline docs/source/examples/ipc.md:7]: Is "better performance" verified experimentally?

Comment (n063h) [reply docs/source/examples/ipc.md:7]: Yes, based on the benchmark experiment at https://github.com/mosecorg/numbin/blob/master/benchmark/bench.py, the results show that when comparing the performance ,Redis connected via a Unix socket > Plasma > Redis connected via a TCP socket.

File: mosec/plugins/redis_shm.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+sent via the original way.
+
+    use case: large image tensors, cluster-shared data
+    benefits: more stable P99 latency, consistency
+"""
+
+from functools import partial
+from typing import Callable, List, Union
+
+import redis
```
[the remark this answers is not in the document]
Comment (n063h) [reply mosec/plugins/redis_shm.py:27]: sry, I havn't imported deps before, overlooked it.

File: docs/source/examples/ipc.md (hunk 1 above)
Comment (lkevinzc) [reply docs/source/examples/ipc.md]: Cool! 

But the benchmark file mentioned above only tests the serialization / de-serialization performance, so you modify it to test the shared memory performance?

Comment (n063h) [reply docs/source/examples/ipc.md]: Since it's a repo of serde library, I reused the cases while did not update it.

Comment (kemingy) [reply docs/source/examples/ipc.md]: I think the code and benchmark data can be uploaded to [mosecorg/benchmark](https://github.com/mosecorg/benchmark). We can add the link here.

File: mosec/plugins/redis_shm.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        """
+        self.client = redis.from_url(url)
+
+        if not id_gen:
+
+            def int_to_bytes(number: int) -> bytes:
+                return str(number).encode()
+
+            def bytes_id_gen() -> bytes:
+                int_id_gen = partial(self.client.incr, DEFAULT_KEY)
```
Comment (kemingy) [inline mosec/plugins/redis_shm.py:62]: If I run multiple mosec instances, will they share the same key to the same redis?

**Comment by kemingy** [comment]
The implementation can refer to the new plasma mixin https://github.com/mosecorg/mosec/blob/main/mosec/mixin/plasma_worker.py

File: mosec/plugins/redis_shm.py (hunk 3 above)
Comment (lkevinzc) [reply mosec/plugins/redis_shm.py:62]: @n063h Could you please add something (e.g., timestamp) to differentiate?

**Comment by n063h** [comment]
Instead of plugin,now redis shm is implemented as a mixin worker, refer to #367

**Pull request closed without merging**